File: csharp/sieve/Sieve.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class Sieve
{
    public static int[] Primes(int limit)
    {
        // var primes = new List<int>();
        // var composites = new HashSet<int>();
        var numbers = new Dictionary<int, bool>();

        if (limit < 2) {
            throw new ArgumentOutOfRangeException("No primes under 2.");
        }

        for (int i = 2; i <= limit; i++) {
            // if (!composites.Contains(i)) {
            if (!numbers.ContainsKey(i)) {
                // primes.Add(i);
                numbers[i] = true;
                for (int j = i + i; j <= limit; j += i) {
                    // composites.Add(j);
                    numbers[j] = false;
                }
            }
        }
        // return primes.ToArray();
        return numbers.Keys.Where(k => numbers[k]).ToArray();
    }
}

File: kotlin/sieve/src/main/kotlin/Sieve.kt
object Sieve {

    fun primesUpTo(upperBound: Int): List<Int> {
        var composites = mutableListOf<Int>()
        var primes = mutableListOf<Int>()
        for (i in 2..upperBound) {
            if (!composites.contains(i)) {
                primes.add(i)
                for (j in i..upperBound step i) {
                    composites.add(j)
                }
            }
        }
        return primes
    }
}
